lib/src/main/java/io/ably/lib/transport/ConnectionManager.java

```diff
@@ -254,43 +254,49 @@ public void run() {
 	 * transport events/notifications
 	 ***************************************/
 
-	void onMessage(ProtocolMessage message) {
-		if(protocolListener != null)
-			protocolListener.onRawMessage(message);
-		switch(message.action) {
-		case HEARTBEAT:
-			onHeartbeat(message);
-			break;
-		case ERROR:
-			ErrorInfo reason = message.error;
-			if(reason == null)
-				Log.e(TAG, "onMessage(): ERROR message received (no error detail)");
-			else
-				Log.e(TAG, "onMessage(): ERROR message received; message = " + reason.message + "; code = " + reason.code);
+	void onMessage(ProtocolMessage message) throws AblyException {
```

Comment:
@gokhanbarisaker: overall, I agree: I don't want to add micro-level validation throughout (especially in the manner of your Monad Validator which is IMO needlessly complex and verbose and also creates unnecessary garbage) in order to validate the input from our own server.

However, even when the client library and server are working perfectly, things can go wrong: for example some cellular networks transcode JSON to HTML (!) if you're not running over TLS. Therefore it is possible that we will get deserialisation failures for messages (this is most likely) and if deserialisation succeeds it is possible we will get protocol errors. These errors don't need to be diagnosed but should at least be handled so we don't end up with exceptions propagating up to kill threads or whatever.

So I think the code in `WsClient` should have a try/catch around the deserialisation; if an error is detected it should be logged as protocol error, and the transport disconnected. Similarly any exception arising in this scope here should I think be caught, and logged as a protocol error with the transport disconnected.

WDYT?